Client/mods/deathmatch/logic/CVehicleUpgrades.cpp

```diff
@@ -64,376 +289,28 @@ bool CVehicleUpgrades::IsUpgradeCompatible(unsigned short usUpgrade)
-        // Initialisation happens when we load the clump which is done when we require a specific model rather than in bulk
-        if (supportedUpgrades.m_bInitialised == true)
-        {
-            // Hood and Vents
-            if (ucSlot == 0 || ucSlot == 1)
-            {
-                if (supportedUpgrades.m_bBonnet == true)
-                {
-                    if (us == 1142 || us == 1144)
-                    {
-                        // We add both left and right for some reason seems to be SA doing it
-                        bReturn &= supportedUpgrades.m_bBonnet_Left;
-                        bReturn &= supportedUpgrades.m_bBonnet_Left_dam;
+    int modelRequestIDParent = g_pGame->GetModelInfo(usModel)->GetParentID();
 
-                        bReturn &= supportedUpgrades.m_bBonnet_Right;
-                        bReturn &= supportedUpgrades.m_bBonnet_Right_dam;
-                    }
-                    else if (us == 1004 || us == 1005 || us == 1011 || us == 1012)
-                    {
-                        // Just needs m_bBonnet
-                        bReturn = true;
-                    }
-                }
-            }
-            else if (ucSlot == 2)
-            {
-                bReturn = supportedUpgrades.m_bSpoiler;
-            }
-            else if (ucSlot == 3)
-            {
-                bReturn = supportedUpgrades.m_bSideSkirt_Right;
-                bReturn &= supportedUpgrades.m_bSideSkirt_Left;
-            }
-            else if (ucSlot == 4)
-            {
-                bReturn = supportedUpgrades.m_bFrontBullbars;
-            }
-            else if (ucSlot == 5)
-            {
-                bReturn = supportedUpgrades.m_bRearBullbars;
-            }
-            else if (ucSlot == 6)
-            {
-                bReturn = supportedUpgrades.m_bLamps;
-                bReturn &= supportedUpgrades.m_bLamps_dam;
-            }
-            else if (ucSlot == 7)
-            {
-                bReturn = supportedUpgrades.m_bRoof;
-            }
-            else if (ucSlot == 13)
-            {
-                bReturn = supportedUpgrades.m_bExhaust;
-            }
-            else if (ucSlot == 14)
-            {
-                bReturn = supportedUpgrades.m_bFrontBumper;
-            }
-            else if (ucSlot == 15)
-            {
-                bReturn = supportedUpgrades.m_bRearBumper;
-            }
-            else if (ucSlot == 16)
-            {
-                bReturn = supportedUpgrades.m_bMisc;
-            }
-        }
-    }
-    return bReturn;
+   if (auto* pModelInfo = g_pGame->GetModelInfo(usModel); pModelInfo && pModelInfo->GetParentID() != 0)
+    usModel = static_cast<uint16_t>(pModelInfo->GetParentID());
+
+    size_t vehicleIndex = (usModel - 400);
+    size_t upgradeIndex = (usUpgrade - 1000);
+
+    const VehicleUpgradeSlots& slots = vehicleUpgradeSlots[vehicleIndex];
```

Comment:
This is unsafe since potentially `usModel`(and `usUpgrade`) can be extented in future and you will get out of the range access problem. Maybe you should use `at` method instead of [] to properly handle with the help of exceptions? Or at least add `assert` for the early diagnosis.